From c62f08852e524ad0b68238d03d3e5232ba9b85bc Mon Sep 17 00:00:00 2001
From: Kaisheng Yang <kaisheng@qti.qualcomm.com>
Date: Wed, 11 Mar 2026 10:45:29 +0800
Subject: [PATCH 2/2] early-ramdisk-init: Move GVM preload unit to late tasklet

Move GVM preload functions to tasklet-lib. init.c only keep main
stand functions. Mount tmpfs before late tasklet run.

Change-Id: I3a3bf3d02000d163cc83b82fb139a963d1b32b3a
Signed-off-by: Kaisheng Yang <kaisheng@qti.qualcomm.com>

Upstream-Status: Inappropriate
---
 conf/gvm-gen4-5/tasklet.late       |   1 +
 conf/quin-gvm-gen4-5/tasklet.late  |   1 +
 conf/quin-tgvm-gen4-5/tasklet.late |   1 +
 src/include/utils.h                |   3 -
 src/init/fast-modules-load.c       |  10 ---
 src/init/init.c                    | 120 +----------------------------
 src/utils/tasklet-lib.c            | 107 ++++++++++++++++++++++++-
 7 files changed, 112 insertions(+), 131 deletions(-)

diff --git a/conf/gvm-gen4-5/tasklet.late b/conf/gvm-gen4-5/tasklet.late
index e69de29..e889f3a 100644
--- a/conf/gvm-gen4-5/tasklet.late
+++ b/conf/gvm-gen4-5/tasklet.late
@@ -0,0 +1 @@
+@video_lib_unification_tasklet
diff --git a/conf/quin-gvm-gen4-5/tasklet.late b/conf/quin-gvm-gen4-5/tasklet.late
index e69de29..e889f3a 100644
--- a/conf/quin-gvm-gen4-5/tasklet.late
+++ b/conf/quin-gvm-gen4-5/tasklet.late
@@ -0,0 +1 @@
+@video_lib_unification_tasklet
diff --git a/conf/quin-tgvm-gen4-5/tasklet.late b/conf/quin-tgvm-gen4-5/tasklet.late
index e69de29..e889f3a 100644
--- a/conf/quin-tgvm-gen4-5/tasklet.late
+++ b/conf/quin-tgvm-gen4-5/tasklet.late
@@ -0,0 +1 @@
+@video_lib_unification_tasklet
diff --git a/src/include/utils.h b/src/include/utils.h
index 78dec6d..24d075d 100644
--- a/src/include/utils.h
+++ b/src/include/utils.h
@@ -75,7 +75,4 @@ void inline safe_close(int fd)
 	return;
 }
 
-extern char machine_name[128];
-extern int soc_id;
-
 #endif
diff --git a/src/init/fast-modules-load.c b/src/init/fast-modules-load.c
index 0a2ae3a..ec4eb9d 100644
--- a/src/init/fast-modules-load.c
+++ b/src/init/fast-modules-load.c
@@ -457,18 +457,8 @@ static int late_tasklet_run(char *task, int load_mode)
 	else if(pid == 0) {
 		log_info("Run tasklet: %s\n", task_name);
 
-		if(ret = unshare(CLONE_NEWNS)){
-			log_kmsg("Tasklet %s unshare failed: %d", task_name, errno);
-			exit(ret);
-		}
-		if(ret = mount_setup()){
-			log_kmsg("Tasklet %s tmpfs setup failed: %d", task_name, ret);
-			return ret;
-		}
-
 		ret = func(NULL);
 
-		mount_unsetup();
 		exit(ret);
 	}
 
diff --git a/src/init/init.c b/src/init/init.c
index 5bf6ff1..c0e4727 100644
--- a/src/init/init.c
+++ b/src/init/init.c
@@ -430,66 +430,6 @@ void mount_unsetup(void)
 	umount("/dev");
 }
 
-#ifdef LIB_UNIFICATION
-static int uni_overlayfs(char* path, char* machine, const char* chip)
-{
-	int ret = 0;
-	char lower_dir[128];
-
-	if (machine != NULL && machine[0] == '\0')
-		snprintf(lower_dir, sizeof(lower_dir), "lowerdir=/uni/%s%s:%s", chip, path, path);
-	else if (chip != NULL && chip[0] == '\0') {
-		if (!strcmp(machine_name, "SA_QX_VM"))
-			snprintf(lower_dir, sizeof(lower_dir), "lowerdir=/uni/hqx%s:%s", path, path);
-		else if (!strcmp(machine_name, "SA_GUNYAH_VM"))
-			snprintf(lower_dir, sizeof(lower_dir), "lowerdir=/uni/hgy%s:%s", path, path);
-	} else {
-		if (!strcmp(machine_name, "SA_QX_VM"))
-			snprintf(lower_dir, sizeof(lower_dir), "lowerdir=/uni/hqx/%s%s:%s", chip, path, path);
-		else if (!strcmp(machine_name, "SA_GUNYAH_VM"))
-			snprintf(lower_dir, sizeof(lower_dir), "lowerdir=/uni/hgy/%s%s:%s", chip, path, path);
-	}
-
-	ret = mount("overlay", path, "overlay", MS_NOATIME, lower_dir);
-	if(0 != ret) {
-		log_kmsg("mount %s error: %d with lowerdir: %s\n", path, errno, lower_dir);
-	} else {
-		log_kmsg("mount %s overlayfs %s done\n", lower_dir, path);
-	}
-
-	return ret;
-}
-
-static int uni_bindfs(char* path, char* machine, char* chip)
-{
-	int ret = 0;
-	char lower_dir[128];
-
-	if (machine != NULL && machine[0] == '\0')
-		snprintf(lower_dir, sizeof(lower_dir), "lowerdir=/uni/%s%s", chip, path);
-	else if (chip != NULL && chip[0] == '\0') {
-		if (!strcmp(machine_name, "SA_QX_VM"))
-			snprintf(lower_dir, sizeof(lower_dir), "/uni/hqx%s", path);
-		else if (!strcmp(machine_name, "SA_GUNYAH_VM"))
-			snprintf(lower_dir, sizeof(lower_dir), "/uni/hgy%s", path);
-	} else {
-		if (!strcmp(machine_name, "SA_QX_VM"))
-			snprintf(lower_dir, sizeof(lower_dir), "/uni/hqx/%s%s", chip, path);
-		else if (!strcmp(machine_name, "SA_GUNYAH_VM"))
-			snprintf(lower_dir, sizeof(lower_dir), "/uni/hgy/%s%s", chip, path);
-	}
-
-	ret = mount(lower_dir, path, NULL, MS_BIND, NULL);
-	if(0 != ret) {
-		log_kmsg("bind %s error: %d with lowerdir: %s\n", path, errno, lower_dir);
-	} else {
-		log_kmsg("bind %s overlayfs %s done\n", lower_dir, path);
-	}
-
-	return ret;
-}
-#endif
-
 int main(int argc, char* argv[])
 {
 	int ret;
@@ -591,63 +531,10 @@ int main(int argc, char* argv[])
 		return -1;
 	}
 
-	late_tasklet_load(cmd.mode);
-
-#ifdef LIB_UNIFICATION
-	pid = fork();
-	if(pid < 0) {
-		log_kmsg("error: fork process for lib unification fail\n");
-		if(mount("sysfs", "/sys", "sysfs", MS_SILENT, NULL)) {
-			log_kmsg("mount sysfs failed: %d\n", errno);
-		}
-		ret = mount("overlay", "/etc", "overlay", MS_NOATIME, "lowerdir=/etc:/uni/lemans/etc");
-		if(ret == -1) {
-			log_kmsg("mount lowerdir=/etc:/uni/lemans/etc error: %d\n", errno);
-		}
-		if(umount("/sys")) {
-			log_kmsg("umount sysfs failed: %d\n", errno);
-		}
-	}
-	else if(pid == 0) {
-		const char *socid_name;
-		/* translate socid_name */
-		switch (soc_id) {
-			case 532:
-			case 533:
-			case 534:
-			case 535:
-				socid_name = "lemans";
-				break;
-			case 606:
-			case 695:
-				socid_name = "monaco";
-				break;
-			default:
-				socid_name = NULL;
-				break;
-		}
-
-		// Handle overlay
-		if (!strcmp(machine_name, "SA_QX_VM") || !strcmp(machine_name, "SA_GUNYAH_VM")) {
-			//video overlayfs
-			uni_overlayfs("/usr/lib", "", socid_name);
-			uni_overlayfs("/etc", "", socid_name);
-
-			// security overlay
-			uni_overlayfs("/usr/bin", machine_name, "");
-			uni_overlayfs("/usr/lib", machine_name, "");
-
-			// common overlay
-			//uni_overlayfs("/usr/bin", machine_name, socid_name);
-			//uni_overlayfs("/usr/lib", machine_name, socid_name);
-			//uni_overlayfs("/etc", machine_name, socid_name);
+	if(ret = mount_setup())
+		log_kmsg("Realroot tmpfs setup failed: %d\n", ret);
 
-			// security driver load
-			uni_bindfs("/etc/modules-load.d/security_load.conf", machine_name, "");
-			uni_bindfs("/etc/fstab", machine_name, "");
-		}
-	}
-#endif
+	late_tasklet_load(cmd.mode);
 
 	log_close();
 
@@ -655,6 +542,7 @@ int main(int argc, char* argv[])
 		return errno;
 	}
 
+	mount_unsetup();
 	return 0;
 }
 
diff --git a/src/utils/tasklet-lib.c b/src/utils/tasklet-lib.c
index 2b8f448..343ddac 100644
--- a/src/utils/tasklet-lib.c
+++ b/src/utils/tasklet-lib.c
@@ -219,8 +219,8 @@ retry:
 TASKLET_LATE_CALL("early_mount_tasklet", early_mount_func);
 
 // socinfo information
-char machine_name[128] = {0};
-int soc_id = -1;
+static char machine_name[128] = {0};
+static int soc_id = -1;
 
 int fetch_socinfo(void *data)
 {
@@ -258,3 +258,106 @@ int fetch_socinfo(void *data)
 	return 0;
 }
 TASKLET_EARLY_CALL("fetch_socinfo_tasklet", fetch_socinfo);
+
+static int uni_overlayfs(char* path, char* machine, const char* chip)
+{
+	int ret = 0;
+	char lower_dir[128];
+
+	if (machine != NULL && machine[0] == '\0')
+		snprintf(lower_dir, sizeof(lower_dir), "lowerdir=/uni/%s%s:%s", chip, path, path);
+	else if (chip != NULL && chip[0] == '\0') {
+		if (!strcmp(machine_name, "SA_QX_VM"))
+			snprintf(lower_dir, sizeof(lower_dir), "lowerdir=/uni/hqx%s:%s", path, path);
+		else if (!strcmp(machine_name, "SA_GUNYAH_VM"))
+			snprintf(lower_dir, sizeof(lower_dir), "lowerdir=/uni/hgy%s:%s", path, path);
+	} else {
+		if (!strcmp(machine_name, "SA_QX_VM"))
+			snprintf(lower_dir, sizeof(lower_dir), "lowerdir=/uni/hqx/%s%s:%s", chip, path, path);
+		else if (!strcmp(machine_name, "SA_GUNYAH_VM"))
+			snprintf(lower_dir, sizeof(lower_dir), "lowerdir=/uni/hgy/%s%s:%s", chip, path, path);
+	}
+
+	ret = mount("overlay", path, "overlay", MS_NOATIME, lower_dir);
+	if(0 != ret) {
+		log_kmsg("mount %s error: %d with lowerdir: %s\n", path, errno, lower_dir);
+	} else {
+		log_kmsg("mount %s overlayfs %s done\n", lower_dir, path);
+	}
+
+	return ret;
+}
+
+static int uni_bindfs(char* path, char* machine, char* chip)
+{
+	int ret = 0;
+	char lower_dir[128];
+
+	if (machine != NULL && machine[0] == '\0')
+		snprintf(lower_dir, sizeof(lower_dir), "lowerdir=/uni/%s%s", chip, path);
+	else if (chip != NULL && chip[0] == '\0') {
+		if (!strcmp(machine_name, "SA_QX_VM"))
+			snprintf(lower_dir, sizeof(lower_dir), "/uni/hqx%s", path);
+		else if (!strcmp(machine_name, "SA_GUNYAH_VM"))
+			snprintf(lower_dir, sizeof(lower_dir), "/uni/hgy%s", path);
+	} else {
+		if (!strcmp(machine_name, "SA_QX_VM"))
+			snprintf(lower_dir, sizeof(lower_dir), "/uni/hqx/%s%s", chip, path);
+		else if (!strcmp(machine_name, "SA_GUNYAH_VM"))
+			snprintf(lower_dir, sizeof(lower_dir), "/uni/hgy/%s%s", chip, path);
+	}
+
+	ret = mount(lower_dir, path, NULL, MS_BIND, NULL);
+	if(0 != ret) {
+		log_kmsg("bind %s error: %d with lowerdir: %s\n", path, errno, lower_dir);
+	} else {
+		log_kmsg("bind %s overlayfs %s done\n", lower_dir, path);
+	}
+
+	return ret;
+}
+
+static int video_lib_unification_func(void *data)
+{
+	int ret;
+	char lower_dir[128];
+	const char *socid_name;
+	switch (soc_id) {
+		case 532:
+		case 533:
+		case 534:
+		case 535:
+			socid_name = "lemans";
+			break;
+		case 606:
+		case 695:
+			socid_name = "monaco";
+			break;
+			default:
+			socid_name = NULL;
+			break;
+	}
+
+	// Handle overlay
+	if (!strcmp(machine_name, "SA_QX_VM") || !strcmp(machine_name, "SA_GUNYAH_VM")) {
+		//video overlayfs
+		uni_overlayfs("/usr/lib", "", socid_name);
+		uni_overlayfs("/etc", "", socid_name);
+
+		// security overlay
+		uni_overlayfs("/usr/bin", machine_name, "");
+		uni_overlayfs("/usr/lib", machine_name, "");
+
+		// common overlay
+		//uni_overlayfs("/usr/bin", machine_name, socid_name);
+		//uni_overlayfs("/usr/lib", machine_name, socid_name);
+		//uni_overlayfs("/etc", machine_name, socid_name);
+
+		// security driver load
+		uni_bindfs("/etc/modules-load.d/security_load.conf", machine_name, "");
+		uni_bindfs("/etc/fstab", machine_name, "");
+	}
+
+	return 0;
+}
+TASKLET_LATE_CALL("video_lib_unification_tasklet", video_lib_unification_func);
-- 
2.34.1

